lib/concurrent/timer_task2.rb

```diff
@@ -0,0 +1,325 @@
+  #   task.shutdown
+  #
+  #   task = Concurrent::TimerTask2.new(execution_interval: 1, channel: channel){ raise StandardError }
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel).wait
+  #   end
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   task.shutdown
+  #   channel.pop
+  #
+  # @see http://docs.oracle.com/javase/7/docs/api/java/util/TimerTask.html
+  class TimerTask2 < ::Concurrent::Synchronization::Object
+    safe_initialization!
+
+    # Default `:execution_interval` in seconds.
+    EXECUTION_INTERVAL = 60
+    # Default `:timeout_interval` in seconds.
+    TIMEOUT_INTERVAL = 30
+
+    # Create a new TimerTask2 with the given task and configuration.
+    #
+    # @!macro timer_task_initialize
+    #   @param [Hash] opts the options defining task execution.
+    #   @option opts [Integer] :execution_interval number of seconds between
+    #     task executions (default: EXECUTION_INTERVAL)
+    #   @option opts [Integer] :timeout_interval number of seconds a task can
+    #     run before it is considered to have failed (default: TIMEOUT_INTERVAL)
+    #   @option opts [Boolean] :run_now Whether to run the task immediately
+    #     upon instantiation or to wait until the first execution_interval
+    #     has passed (default: false)
+    #   @option opts [Symbol] :reschedule when to schedule next execution relative,
+    #     relative to a current one, `:before` or `:after` (default: :after)
+    #   @option opts [Promises::Channel, nil] :channel if given, execution results
+    #     will be pushed into the channel (default: nil)
+    #
+    #   @raise ArgumentError when no block is given.
+    #
+    #   @yield to the block after :execution_interval seconds have passed since
+    #     the last yield
+    #   @yieldparam task a reference to the `TimerTask` instance so that the
+    #     block can control its own lifecycle. Necessary since `self` will
+    #     refer to the execution context of the block rather than the running
+    #     `TimerTask`.
+    #   @yieldparam cancellation a cancellation object representing the joined cancellation
+    #     of the timer task and this run's timeout
+    #
+    #   @return [TimerTask] the new `TimerTask`
+    def initialize(opts = {}, &task)
+      raise ArgumentError.new('no block given') unless block_given?
+      raise ArgumentError.new('reschedule must be either :before or :after') unless [nil, :before, :after].include?(opts[:reschedule])
+      @ExecutionInterval = AtomicReference.new nil
+      @TimeoutInterval = AtomicReference.new nil
+      @Cancellation = AtomicReference.new nil
+
+      self.execution_interval = opts[:execution] || opts[:execution_interval] || EXECUTION_INTERVAL
+      self.timeout_interval = opts[:timeout] || opts[:timeout_interval] || TIMEOUT_INTERVAL
+
+      @Channel = opts[:channel]
```

Comment:
I think we discussed some problems which could arise if we allow the channel to be shared between two timer-tasks. Like the arriving messages would not have clear sender. Is that right? It would be better if it would accept arguments for creating the channel, where new channel would be created on dup, or to add dup support on the channel and dup it as well in timer-task's dup.

Reply:
Right, how about we drop [the reuse of channel in #dup](https://github.com/ruby-concurrency/concurrent-ruby/pull/829/files#diff-5d2534adbdf5de2a2ec965af44ecc206R200). This way `#dup`ed timers won't have a channel, unless a new channel is provided when `#dup` is called?

Reply:
That would work but I think it would more surprising. Assuming that a duped timer-task is most often used in the same place/functional-context as the template than it would be expected the new timer-task to have a channel as the parent has.